### ISSUE TITLE
`object-curly-newline`: new option request

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#6488]
I'm on eslint v2.13.1.

The `multiline: true` option is excellent: it lets me require that when newlines exist, every property must be on its own line.

Combining it with `minProperties` is half useful, however: when it's set to, say, 3, that means that any objects with 3 or more properties _must_ be split onto multiple lines, which is great! Unfortunately, this also means that objects with less than 3 properties _must not_ be split onto multiple lines.

What I'd like instead is something like `{ multiline: true, maxPropertiesBeforeForcedNewline: 2 }` which says that if there are `> 2` properties, the object _must_ be multiline, but if there are `<= 2`, the object may be single or multi, provided that it's consistent with the `multiline: true` option.

**Comment by eslintbot** [comment]
Hi @ljharb, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by kaicataldo** [comment]
Thanks for contributing! Sorry if I'm misunderstanding what you're requesting, but I believe the current implementation does what you're asking for.

With this config:
```json
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module",
  },
  "rules": {
    "object-curly-newline": ["error", { "multiline": true, "minProperties": 4 }]
  }
}
```
This warns:
```js
const a = { b, c, d, e };
```

and these don't:
```js
const a = { b, c, d };

const a = {
  b,
  c,
  d
};
```

That being said, the docs are incorrect
> {minProperties: <integer>} - requires line breaks if the number of properties is more than the given integer. Otherwise, disallows line breaks.

should really say
>requires line breaks if the number of properties is equal to or greater than the given integer.

**Comment by ljharb** [comment]
@kaicataldo that is indeed what I'd like it to do - but when I put the following in a file:
```js
var a = { a: 1, b: 2 };
var b = {
  a: 1,
  b: 2
};

 var c = Object.assign({}, a, b, {
   a: 3
 });
```
and then run `eslint --stdin --rule 'object-curly-newline:[2,{"multiline":true,"minProperties":3}]'`, I get:
```shell
<text>
  7:34  error  Unexpected a line break after this open brace    object-curly-newline
  9:2   error  Unexpected a line break before this close brace  object-curly-newline
```

Why is it warning on the last example, when the only requirement here is that 3 or more properties forces a newline?

**Comment by mysticatea** [comment]
Thank you for this issue.

Maybe similar to #6369.
I think this is reasonable.

What do you think `either` option?

```js
/*eslint object-curly-newline: [error, {either: true}]*/

// ✘ BAD
var obj = {a: 1
}
var obj = {
    a: 1}

// ✔ GOOD
var obj = {a: 1}
var obj = {
    a: 1
}
```

- `"always"` - requires line breaks always.
- `"never"` - disallows line breaks.
- `{multiline: true}` (default) - requires line breaks if there are line breaks inside properties or between properties. Otherwise, disallows line breaks.
- `{minProperties: <integer>}` - requires line breaks if the number of properties is more than the given integer. Otherwise, disallows line breaks.
- `{either: true}` - requires line breaks (of both braces) if either brace has a line break.

`multiline` and `minProperties` and `either` can be combined.

- `{multiline: true, minProperties: <integer>, either: true}` - requires line breaks if there are line breaks inside properties or between properties, or if the number of properties is more than the given integer, or if either brace has a line break. Otherwise, disallows line breaks.

If `either` option is enabled, this rule never disallows line breaks.

**Comment by ljharb** [comment]
Sounds great, but the problem is the "Otherwise, disallows line breaks" for "minProperties" :-/ I want a number - say, 3 - above which it *must* be multiline, but below which it can be either multiline or single-line, provided that newlines are consistently applied. Your proposal doesn't seem to allow for that?

**Comment by mysticatea** [comment]
Maybe I mistook wording.
My intention of `{either: true}` is that either or both of braces have line breaks, requires line breaks.

In `{multiline: true, minProperties: 3, either: true}` case, if one of the following conditions satisfied, requires line breaks.

- There are line breaks inside properties or between properties
- The number of properties is more than 3
- Either or both of braces have line break(s)

The following code satisfies the 3rd condition.

```js
var obj = {
    a: 1
}
```

**Comment by ljharb** [comment]
In that case I think it would work perfectly :-)

**Comment by lukeapage** [comment]
Fantastic, I'm getting the exact same problem - I was confused because the existing documentation for minProperties says `requires line breaks if the number of properties is more than the given integer` when it should have that extra text you added `Otherwise, disallows line breaks.`

IMO. `either` is quite a generic term and makes me think you are essentially switching off this rule, rather than relaxing the disallow on minProperties. (You can see that by the fact @ljharb got confused even though you were adding the option he asked for).

Some alternative suggestions ;-
* `maxProperties` - requires line breaks if the number of properties is more than the given integer. 
* Deprecate the confusingly named minProperties and have `requireMinProperties` - number of properties before the new line is required and `disallowMaxProperties` - number of properties at which new line is not allowed

but either way, object-curly-newline is useless to me until an option is added, so am happy with any way to solve the problem.

**Comment by mysticatea** [comment]
I'd like to forward https://github.com/eslint/eslint/issues/6488#issuecomment-227383081.
This proposal allows people to make readable line breaks anytime. (also, this maybe resolve #6369 and #6549).

@eslint/eslint-team May I be given :+1:s, or other thoughts?

**Comment by BYK** [comment]
I'm in! :)

**Comment by mysticatea** [comment]
@lukeapage I'm sorry for I have missed you.

I designed this option as "the condition that it requires line breaks."
Granted, it may be confusing...
I'm interesting to investigate more proper options.

Separating `disallow` and `require` in options sounds good to me.
Maybe...:

```js
{
    "disallowNewline": {
        "singleline": true,
        "maxProperties": 1,
    },
    "requireNewline": {
        "multiline": true,
        "minProperties": 2,
    }
}
```

**Comment by nzakas** [comment]
We have stayed away from having both "disallow" and ` require" in the same rule because it's a bit confusing. Let's try to come up with something a bit clearer.

**Comment by jfelchner** [comment]
@nzakas I think it makes sense to do two things:

* `consistent` as a boolean which makes sure that if one curly has a newline, both curly's have newlines (and vice versa)
* `minProperties` needs to have a bug fixed.  The minimum number of properties required before the rule enables itself.  Right now if you set `minProperties` to `1`, the rule (with or without `multiline` set to `true`, still fails this case:

```javascript
{
  "foo": "bar"
}
```

And considering the docs say:

> requires line breaks if the number of properties is more than the given integer

that seems like a bug.

So to get back to the example above:

```javascript
var c = Object.assign({}, a, b, {
   a: 3
 });
```

That would be solved with simply `minProperties` being `1`.  If you wanted to allow either:

```javascript
var c = Object.assign({}, a, b, {
  a: 3
});

// or

var c = Object.assign({}, a, b, {a: 3});

// but not

var c = Object.assign({}, a, b, {a: 3
});

// or

var c = Object.assign({}, a, b, {
  a: 3});
```

Then you could set `minProperties` to `2` and `consistent` to `true`.

**Comment by jfelchner** [comment]
I'm not sure, however, if this will fix the fact that `multiline` doesn't consider an object with a single property on its own line to be multiline.  This is a separate thing that seems like a bug.

**Comment by nzakas** [comment]
It seems like this conversation has strayed a bit from the original idea. Can someone please summarize the current proposal(s) using the [template](https://github.com/eslint/eslint/blob/master/templates/rule-change-proposal.md)

**Comment by ljharb** [comment]
**What rule do you want to change?**
`object-curly-newline`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option: `consistent`. When `true`, it requires that either both curly braces, or neither, directly enclose newlines.

**Please provide some example code that this change will affect:**
```sh
cat temp
```
```js
var one = { a: 1 };

var two = { a: 1, b: 2 };

var three = { a: 1, b: 2, c: 3 };

var oneM = {
  a: 1
};

var twoM = {
  a: 1,
  b: 2
};

var threeM = {
  a: 1,
  b: 2,
  c: 3
};

var wat = { a: 1
};

var wat2 = {
a: 1 };

var wat3 = { a: 1,
  b: 2,
  c: 3 };
```

**What does the rule currently do for this code?**
```sh
cat temp | ./node_modules/.bin/eslint --stdin --no-eslintrc --rule='object-curly-newline:[2, {"minProperties": 3, "multiline": true}]'

<text>
   5:13  error  Expected a line break after this opening brace   object-curly-newline
   5:32  error  Expected a line break before this closing brace  object-curly-newline
   7:12  error  Unexpected line break after this opening brace   object-curly-newline
   9:1   error  Unexpected line break before this closing brace  object-curly-newline
  23:1   error  Unexpected line break before this closing brace  object-curly-newline
  25:12  error  Unexpected line break after this opening brace   object-curly-newline
  28:12  error  Expected a line break after this opening brace   object-curly-newline
  30:8   error  Expected a line break before this closing brace  object-curly-newline

✖ 8 problems (8 errors, 0 warnings)
```

**What will the rule do after it's changed?**
```sh
cat temp | ./node_modules/.bin/eslint --stdin --no-eslintrc --rule='object-curly-newline:[2, {"minProperties": 3, "multiline": true, "consistent": true}]'

<text>
   5:13  error  Expected a line break after this opening brace   object-curly-newline
   5:32  error  Expected a line break before this closing brace  object-curly-newline
  23:1   error  Unexpected line break before this closing brace  object-curly-newline
  25:12  error  Unexpected line break after this opening brace   object-curly-newline
  28:12  error  Expected a line break after this opening brace   object-curly-newline
  30:8   error  Expected a line break before this closing brace  object-curly-newline

✖ 6 problems (6 errors, 0 warnings)
```

Perhaps a better alternative is actually fixing `minProperties` to mean "min properties to force multiline", instead of meaning "number of properties that determines whether it must be multiline or single-line", but that gets into the require/disallow complexity.
